sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Blob/ContainerRegistryBlobClient.cs

```diff
@@ -740,22 +739,140 @@ public virtual async Task<Response<DownloadBlobResult>> DownloadBlobAsync(string
         }
 
+        /// <summary>
+        /// </summary>
+        /// <param name="digest">The digest of the blob to download.</param>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        /// <returns></returns>
+        public virtual Response<DownloadBlobStreamingResult> DownloadBlobStreaming(string digest, CancellationToken cancellationToken = default)
+        {
+            Argument.AssertNotNull(digest, nameof(digest));
+
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(ContainerRegistryBlobClient)}.{nameof(DownloadBlobStreaming)}");
+            scope.Start();
+            try
+            {
+                return DownloadBlobStreamingInternalAsync(digest, false, cancellationToken).EnsureCompleted();
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="digest">The digest of the blob to download.</param>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        /// <returns></returns>
+        public virtual async Task<Response<DownloadBlobStreamingResult>> DownloadBlobStreamingAsync(string digest, CancellationToken cancellationToken = default)
+        {
+            Argument.AssertNotNull(digest, nameof(digest));
+
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(ContainerRegistryBlobClient)}.{nameof(DownloadBlobStreaming)}");
+            scope.Start();
+            try
+            {
+                return await DownloadBlobStreamingInternalAsync(digest, true, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        private async Task<Response<DownloadBlobStreamingResult>> DownloadBlobStreamingInternalAsync(string digest, bool async, CancellationToken cancellationToken)
+        {
+            ResponseWithHeaders<Stream, ContainerRegistryBlobGetBlobHeaders> blobResult = async ?
+                await _blobRestClient.GetBlobAsync(_repositoryName, digest, cancellationToken).ConfigureAwait(false) :
+                _blobRestClient.GetBlob(_repositoryName, digest, cancellationToken);
+
+            SHA256 sha256 = SHA256.Create();
+            int contentLength = (int)blobResult.Headers.ContentLength.Value;
+            int received = 0;
+            bool validated = false;
+
+            // Wrap the response Content in a RetriableStream so we
+            // can return it before it's finished downloading, but still
+            // allow retrying if it fails.
+            Stream stream = RetriableStream.Create(
+                blobResult.Value,
+                offset => _blobRestClient.GetChunk(_repositoryName, digest, new HttpRange(offset).ToString(), cancellationToken).Value,
```

Comment:
If it isn't too late (i.e. we don't use `RetriableStream` in too many places), maybe we should replace two delegates with one that has `async` parameter.

Reply:
Tracking this separately with https://github.com/Azure/azure-sdk-for-net/issues/34992